### PR TITLE
Feature/fm type

### PR DESCRIPTION
**Pull request by Matthew-Kulich**

### CODE REVIEW

File: src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java [hunk 1]
```diff
@@ -382,4 +381,23 @@ protected boolean removeVisited(Behavior b){
...
+            fEvent.setEventType(EventType.BASIC);
+            fEvent.setGateType(GateType.UNUSED);
+            fEvent.setProbability(1.);
+        } else {
+            fEvent.setEventType(EventType.INTERMEDIATE);
+            fEvent.setGateType(GateType.OR);
+        }
+    }
+
+    private boolean isFailureModeCause(Behavior behavior) {
```
Comment (kostobog) [inline src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java:399]: This function seams generic enough to be added to the Behavior entity.

File: src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java [hunk 2]
```diff
@@ -382,4 +381,23 @@ protected boolean removeVisited(Behavior b){
...
+        return behavior instanceof FailureMode
+                && ((FailureMode) behavior).getFailureModeType() == FailureModeType.FailureModeCause;
+    }
```
Comment (kostobog) [inline src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java:402]: This function seams generic enough to be added to the Behavior entity.

File: src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java [hunk 3]
```diff
@@ -207,12 +207,17 @@ private void processBehavior(Behavior behavior, FaultEvent parentFaultEvent) thr
 
         if (!impairingBehaviors.isEmpty()) {
             for (Behavior impairingBehavior : impairingBehaviors) {
-                if(isVisited(impairingBehavior))
-                    continue;
+                if(isVisited(impairingBehavior) || isFailureModeCause(impairingBehavior)) continue;
                 faultEvents.add(processImpairingBehavior(impairingBehavior, parentFaultEvent));
             }
+        }
```
Comment (kostobog) [inline src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java:213]: Code similar to the one in setFaultEventTypes except the condition. If it makes sense implement `setFaultEventTypes(boolean isBasic, FaultEvent fEvent)` and use it here and line 245.

File: src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java [hunk 4]
```diff
@@ -207,12 +207,17 @@ private void processBehavior(Behavior behavior, FaultEvent parentFaultEvent) thr
...
+        if(faultEvents.size() > 0) {
             parentFaultEvent.setEventType(EventType.INTERMEDIATE);
             parentFaultEvent.setGateType(GateType.OR);
+        }else{
+            parentFaultEvent.setEventType(EventType.BASIC);
+            parentFaultEvent.setGateType(GateType.UNUSED);
+            parentFaultEvent.setProbability(1.);
         }
```
Comment (kostobog) [inline src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java:221]: Code similar to the one in setFaultEventTypes except the condition. If it makes sense implement `setFaultEventTypes(boolean isBasic, FaultEvent fEvent)` and use it here and line 245.

Comment (kostobog) [inline src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java:221]: Code similar to the one in setFaultEventTypes except the condition. If it makes sense implement `setFaultEventTypes(boolean isBasic, FaultEvent fEvent)` and use it here and line 245.

File: src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java (hunk 2 above)
Comment (Matthew-Kulich) [reply src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java:402]: Ok

File: src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java (hunk 4 above)
Comment (Matthew-Kulich) [reply src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java:221]: Ok, its makes sense

**Comment by Matthew-Kulich** [comment]
Fixed. @kostobog

**Comment by kostobog** [comment]
I still see some issues with the generated FMEA. Here is an example, try to check it locally and fix the generation accordingly.
Consider the following system description and FTA:
image
 
For this system and FTA I get the following table:
image
The table has the following issues:
- two rows are wrong - indicated by the cross "x" at the end of the row
- one row is missing - indicated by the question mark "?" at the end of the row
- there is one row that has set its cause incorrectly.

Here is how FMEA generation should work with this example system and FTA.
The paths in the FTA tree are:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kostobog/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kostobog/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



tree paths |   |   |   |  
-- | -- | -- | -- | --
path1 | FM1 | F2 fails | F1 fails | F fails
path2 | FM2 | F2 fails | F1 fails | F fails
path3 | FM3 | F2 fails | F1 fails | F fails
path4 |   | FM3 | F1 fails | F fails



</body>

</html>


Processing the paths and considering causes attached to the selected function F or failure mode FM should generate the following list of rows:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kostobog/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kostobog/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



order |   | 2 | 1 | 4 | 3 | 5 | 7 | 6 |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
pocessed path | row # | Component | Function | Cause | Failure Mode | Local Effect | Next Effect | Final Effect | duplicates
path1 | 1 | C1 | F2 | cause1 | FM1 | F1 fails | - | F fails |  
path1 | 2 | C1 | F2 | cause4 | - | F1 fails | - | F fails |  
path2 | 3 | C1 | F2 | cause2 | FM2 | F1 fails | - | F fails |  
path2 | 4 | C1 | F2 | cause3 | FM2 | F1 fails | - | F fails |  
path2 | 5 | C1 | F2 | cause4 | - | F1 fails | - | F fails | 2
path3 | 6 | C1 | F2 | - | FM3 | F1 fails | - | F fails |  
path3 | 7 | C1 | F2 | cause4 | - | F1 fails | - | F fails | 2
path4 | 8 | C1 | F1 | - | FM4 | - | - | F fails |  



</body>

</html>



The final FMEA table contains unique rows. Here is the table from the example after removing the duplicate rows:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kostobog/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kostobog/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



Resulting FMEA after removing duplcate rows |   |   |   |   |  
-- | -- | -- | -- | -- | --
row # | Component | Function | Cause | Failure Mode | Local Effect | Next Effect | Final Effect
1 | C1 | F2 | cause1 | FM1 | F1 fails | - | F fails
2 | C1 | F2 | cause4 | - | F1 fails | - | F fails
3 | C1 | F2 | cause2 | FM2 | F1 fails | - | F fails
4 | C1 | F2 | cause3 | FM2 | F1 fails | - | F fails
6 | C1 | F2 | - | FM3 | F1 fails | - | F fails
8 | C1 | F1 | - | FM4 | - | - | F fails



</body>

</html>

**Review by kostobog — APPROVED** [review]
Looks good to me.

**Pull request merged**